trin-execution/src/era/binary_search.rs

```diff
@@ -0,0 +1,111 @@
+use e2store::{
+    e2store::types::{Entry, Header as E2StoreHeader},
+    era::{get_beacon_fork, CompressedSignedBeaconBlock, Era},
+    utils::get_era_file_download_links,
+};
+use surf::Client;
+
+use super::{
+    constants::FIRST_ERA_EPOCH_WITH_EXECUTION_PAYLOAD,
+    types::ProcessedEra,
+    utils::{download_raw_era, process_era_file},
+};
+
+pub struct EraBinarySearch {}
+
+impl EraBinarySearch {
+    /// For era files there isn't really a good way to know which era file has the block we want,
+    /// well we could download all of them and check that isn't really feasible. We could also
+    /// create a table which maps block numbers to era files, but that would be a lot of work and
+    /// would need to be updated every time a new era file is added. So we will just binary
+    /// search for the era file which contains the block we want. To make it so we don't have to
+    /// download multiple era files because they are very big we will download the first beacon
+    /// block estimate which blocks it contains and narrow our search till we find it.
+    pub async fn find_era_file(
+        http_client: Client,
+        block_number: u64,
+    ) -> anyhow::Result<ProcessedEra> {
+        let era_links = get_era_file_download_links(&http_client).await?;
```

Comment:
nit: we know the first execution block that is in era file, right?
I would do a check that `block_number` is not less than that.